flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/KafkaUtil.java

```diff
@@ -0,0 +1,116 @@
+     * offset.
+     *
+     * <p>This method will fetch the latest offsets for the partitions once and only return records
+     * until that point.
+     *
+     * @param topic to fetch from
+     * @param properties used to configure the created {@link KafkaConsumer}
+     * @param committed determines the mode {@link ConsumerConfig#ISOLATION_LEVEL_CONFIG} with which
+     *     the consumer reads the records.
+     * @return all {@link ConsumerRecord} in the topic
+     * @throws KafkaException
+     */
+    public static List<ConsumerRecord<byte[], byte[]>> drainAllRecordsFromTopic(
+            String topic, Properties properties, boolean committed) throws KafkaException {
+        final Properties consumerConfig = new Properties();
+        consumerConfig.putAll(properties);
+        consumerConfig.put("key.deserializer", ByteArrayDeserializer.class.getName());
+        consumerConfig.put("value.deserializer", ByteArrayDeserializer.class.getName());
+        consumerConfig.put(
+                ConsumerConfig.ISOLATION_LEVEL_CONFIG,
+                committed ? "read_committed" : "read_uncommitted");
+        try (KafkaConsumer<byte[], byte[]> consumer = new KafkaConsumer<>(consumerConfig)) {
+            Map<Integer, TopicPartition> assignments = getAllPartitions(consumer, topic);
+            Map<TopicPartition, Long> endOffsets = consumer.endOffsets(assignments.values());
+            consumer.assign(assignments.values());
+            consumer.seekToBeginning(assignments.values());
+
+            final List<ConsumerRecord<byte[], byte[]>> consumerRecords = new ArrayList<>();
+            while (!assignments.isEmpty()) {
+                consumer.assign(assignments.values());
+                ConsumerRecords<byte[], byte[]> records = consumer.poll(CONSUMER_POLL_DURATION);
+                LOG.info("Fetched {} records from topic {}.", records.count(), topic);
+
+                // Remove partitions from polling which have reached its end.
+                final Iterator<Map.Entry<Integer, TopicPartition>> assignmentIterator =
+                        assignments.entrySet().iterator();
+                while (assignmentIterator.hasNext()) {
+                    final Map.Entry<Integer, TopicPartition> assignment = assignmentIterator.next();
+                    final TopicPartition topicPartition = assignment.getValue();
+                    final long position = consumer.position(topicPartition);
+                    final long endOffset = endOffsets.get(topicPartition);
+                    LOG.info(
+                            "Endoffset {} and current position {} for partition {}",
+                            endOffset,
+                            position,
+                            assignment.getKey());
+                    if (endOffset - position > 0) {
+                        continue;
+                    }
+                    assignmentIterator.remove();
+                }
```

Comment:
We could have a `finishedPartitions` list in each iteration. Then we don't need the Iterator as we would only modify assignments in a batch way at the end of the iteration. Further, we would only call `consumer.assign` on a change.